src/base/components/usb_logger.rs

```diff
@@ -0,0 +1,103 @@
+// Based on https://github.com/DCRichards/arduino-vagrant-rust/blob/7bdd4a5eb15639a78cf7826a95863e21a98eb0bb/apps/plant/app/src/usb_logger.rs
+use {
+    arduino_nano33iot as bsp,
+    bsp::hal::clock::GenericClockController,
+    bsp::hal::usb::UsbBus,
+    bsp::pac::{interrupt, PM, USB},
+    bsp::usb_allocator,
+    core::fmt::{Error, Write},
+    core::result::Result,
+    cortex_m::interrupt::free,
+    cortex_m::peripheral::NVIC,
+    usb_device::bus::UsbBusAllocator,
+    usb_device::prelude::*,
+    usbd_serial::{SerialPort, USB_CLASS_CDC},
+};
+
+macro_rules! from {
+    ($stone:expr) => {
+        base::components::usb_logger::USBLogger::new(
+            $stone.usb,
+            &mut $stone.clock,
+            &mut $stone.pm,
+            $stone.pins.usb_dm.into(),
+            $stone.pins.usb_dp.into(),
+            &mut $stone.nvic,
+        )
+    };
+}
+pub(crate) use from;
+
+static mut USB_ALLOCATOR: Option<UsbBusAllocator<UsbBus>> = None;
+static mut USB_BUS: Option<UsbDevice<UsbBus>> = None;
+static mut USB_SERIAL: Option<SerialPort<UsbBus>> = None;
+
+/// Represents a logger which logs over USB Serial.
+pub struct USBLogger {}
+
+impl USBLogger {
+    /// Get the log instance
+    pub fn new(
```

Comment:
Why even return `Self`, it seems completely unused?

As things stand with the `static mut` this function should really be called `init` or something.

Reply:
Let's rewrite all this file :)